### PR TITLE
decode more utf-8 strings to utf-16, including emoji and 4+ byte utf-16

### PR DESCRIPTION
**Pull request by adipose**

### CODE REVIEW

**Comment by adipose** [comment]
Patch now supports all utf-16, I believe.  DSUtil routines don't work because they use CStringW, which needs a null terminator, but the single codepoint won't always be null terminated.  Either we could copy to another buffer, decode the entire string at once, or use the method I did of decoding a fixed length with no null terminator.

image

image

image

File: src/Subtitles/TextFile.cpp [hunk 1]
```diff
@@ -588,36 +588,14 @@ BOOL CTextFile::ReadString(CStringW& str)
                     m_wbuffer[nCharsRead] = m_buffer[m_posInBuffer] & 0x7f;
                 } else if (Utf8::isFirstOfMultibyte(m_buffer[m_posInBuffer])) {
                     int nContinuationBytes = Utf8::continuationBytes(m_buffer[m_posInBuffer]);
-                    ASSERT(bValid);
-
-                    if (m_posInBuffer + nContinuationBytes >= m_nInBuffer) {
```
Comment (clsid2) [inline src/Subtitles/TextFile.cpp:593]: Shouldn't this check still be needed, in case there is data for a partial character at the end of the file?

Comment (adipose) [reply src/Subtitles/TextFile.cpp:593]: Yes, I need to add that check, and maybe one other to validate the UTF8 bytes.

**Comment by adipose** [comment]
Screenshots were from this srt file, by the way:

```
1
00:00:01,000 --> 00:00:02,500
𩸽aa (ok)

2
00:00:03,000 --> 00:00:04,500
bb七七七七七 (English & Chinese , good)

3
00:00:05,000 --> 00:00:06,500
cc七七七七七 (both good , again)

4
00:00:07,000 --> 00:00:08,500
dd😁 (Start wrong , Not support Emoji ?)

5
00:00:09,000 --> 00:00:10,500
ee七七七七七 (Chinese , wrong ! This is the bug !)

6
00:00:11,000 --> 00:00:12,500
ff end (English , good)
```

**Comment by k9fvm** [comment]
It should not appear unless you add it manually.
If you added it on purpose, there is no problem.

**Comment by adipose** [comment]
Yes I added it to test non emoji above 0xffff.

**Pull request merged**